### ISSUE TITLE
Actions/predicates see invisible labels

### ISSUE THREAD

**Issue by Mingun** [pegjs/pegjs#396]
Version 0.9.0. Grammar
```pegjs
start
  = visible1:.
    (invisible:.)
    (visible2:. {return [visible1, invisible, visible2]; });
```
must throw js error because variable `invisible` must not be defined, but instead it parses `012` as:
```json
[
   "0",
   "1",
   ["0", "1", "2"]
]
```
Now this not looks dangerous... but wait! if you made small change (rename `invisible` to `visible1`):
```pegjs
start
  = visible1:.
    (visible1:.)
    (visible2:. {return [visible1, visible2]; });
```
...result is (on input `012`)
```json
[
   "0",
   "1",
   ["1", "2"]// hmm...
]
```
----
Second example. Grammar
```pegjs
start
  = (invisible:. .)
    visible1:.
    (visible2:. {return [invisible, visible1, visible2]; });
```
must throw js error because variable `invisible` must not be defined, but instead it parses `0123` as:
```json
[
   ["0", "1"],
   "2",
   ["2", "2", "3"]// hmm...
]
```
Not that you expect it.

**Comment by Mingun** [comment]
By the way, [I prevented](https://github.com/pegjs/pegjs/commit/fb5f6c6ee94b962c45f591f64b293bc11ba57ae6#commitcomment-9743269) that tests for negative cases are necessary. The code block shall or to see the correct values always, or not to see them at all.

**Comment by dmajda** [comment]
The issue here really is that parens do not provide label scope isolation in some cases, specifically when they contain a labeled expression or a sequence. Minimal examples are:

```
start = (a:"a") { return a; }
```

```
start = ("a" b:"b" "c") { return b; }
```

In 0.9.0, these both pass (with appropriate inputs).

---

Now, the question is: Is this behavior a bug or not? To decide that, one needs to consider the purpose of label scope isolation, which is to avoid situations like this:

```
start = (a:"a")* { return a; }
```

Here, it’s unclear what should be the value of `a`. Maybe it should be `"a"` (the last match)? Maybe it should be an array of `"a"`’s (because of `*`)? Moreover, consider cases like this:

```
start = (a:"a" { return a; })* { return a; }
```

If we go the “array of `"a"`’s” way above, then `a` should logically have different value in the two actions (`"a"` in the first one, and array of `"a"`’s in the second one). This is undesirable and confusing.

Label scope isolation avoids these situations by making labels inside operators like `*` invisible to the outside. For consistency, this rule is enforced for all operators, even for ones like `&` where no weird situations should arise (I think).

What all operators which isolate label scope have in common is that when a label is used inside them, the operator’s expression needs to be wrapped in parens. This is necessary to avoid parsing e.g. `a:"a"*` as `a:("a"*)`. **The result is that it *appears* that is’s the parens which introduce scope isolation.** But it’s really the operators. Parens only enforce correct precedence and otherwise they don’t have any semantic meaning.

From that perspective, the current behavior is correct. But it’s highly unintuitive for users, which can easily be led to thinking that it’s parens which limit label scope and then be surprised when that doesn’t happen in some cases. So, my conclusion is: **Yes, this is a valid bug.** We really want to make it so that parens limit label scope.

---

I looked briefly at your fix in #418 and then I decided to take [another route](https://github.com/pegjs/pegjs/commit/0c39f1cf86d73c85a1f832e6cfddf4176f9e58bc) to fixing the problem, introducing a new `group` AST node. In my opinion, that solution is cleaner. I also [refactored the relevant tests a bit](https://github.com/pegjs/pegjs/compare/f07ab7f32e64a697711567983a546a15c379733e...72293186715b1e7cc8bfc606749f5b3a48f01c17) and [introduced some negative test cases](https://github.com/pegjs/pegjs/compare/72293186715b1e7cc8bfc606749f5b3a48f01c17...58806a3d778c784399c9fe831b92a99f9fe267c0) (I think you were right about the need for them after all).

With that, I’m closing this issue.
